google-cloud-firestore/src/test/java/com/google/cloud/firestore/PartitionQuery.java

```diff
@@ -0,0 +1,204 @@
+  public static final Cursor CURSOR1 =
+      Cursor.newBuilder()
+          .addValues(Value.newBuilder().setReferenceValue(DATABASE_NAME + "/collection/doc1"))
+          .build();
+  public static final Cursor PARTITION1 = CURSOR1.toBuilder().setBefore(true).build();
+  public static final Cursor CURSOR2 =
+      Cursor.newBuilder()
+          .addValues(Value.newBuilder().setReferenceValue(DATABASE_NAME + "/collection/doc2"))
+          .build();
+  public static final Cursor PARTITION2 = CURSOR2.toBuilder().setBefore(true).build();
+
+  @Spy
+  private final FirestoreImpl firestoreMock =
+      new FirestoreImpl(
+          FirestoreOptions.newBuilder().setProjectId("test-project").build(),
+          Mockito.mock(FirestoreRpc.class));
+
+  @Mock private UnaryCallable<PartitionQueryRequest, PartitionQueryPagedResponse> callable;
+  @Mock private PartitionQueryPagedResponse pagedResponse;
+  @Mock private PartitionQueryPage queryPage;
+
+  @Captor private ArgumentCaptor<RunQueryRequest> runQuery;
+  @Captor private ArgumentCaptor<ApiStreamObserver> streamObserverCapture;
+  @Captor private ArgumentCaptor<PartitionQueryRequest> requestCaptor;
+
+  @Test
+  public void requestsOneLessThanDesired() throws Exception {
+    int desiredPartitionsCount = 2;
+
+    PartitionQueryRequest expectedRequest =
+        PartitionQueryRequest.newBuilder()
+            .setParent(DATABASE_NAME)
+            .setStructuredQuery(
+                StructuredQuery.newBuilder()
+                    .addFrom(
+                        StructuredQuery.CollectionSelector.newBuilder()
+                            .setAllDescendants(true)
+                            .setCollectionId("collectionId"))
+                    .addOrderBy(
+                        StructuredQuery.Order.newBuilder()
+                            .setField(
+                                StructuredQuery.FieldReference.newBuilder()
+                                    .setFieldPath("__name__"))
+                            .setDirection(StructuredQuery.Direction.ASCENDING)))
+            .setPartitionCount(desiredPartitionsCount - 1)
+            .build();
+
+    PartitionQueryResponse response =
+        PartitionQueryResponse.newBuilder().addPartitions(CURSOR1).build();
+
+    when(pagedResponse.iterateAll()).thenReturn(ImmutableList.of(CURSOR1));
+    when(queryPage.getResponse()).thenReturn(response);
+    doReturn(ApiFutures.immediateFuture(pagedResponse))
+        .when(firestoreMock)
+        .sendRequest(
+            requestCaptor.capture(),
+            Matchers.<UnaryCallable<PartitionQueryRequest, PartitionQueryPagedResponse>>any());
+
+    firestoreMock.collectionGroup("collectionId").getPartitions(desiredPartitionsCount).get();
+
+    PartitionQueryRequest actualRequest = requestCaptor.getValue();
+    assertEquals(actualRequest, expectedRequest);
+  }
+
+  @Test
+  public void doesNotIssueRpcIfOnlyASinglePartitionIsRequested() throws Exception {
```

Comment:
This seems like an interesting optimization. And also like an unlikely case? Is it worth special handling in libs or should we just let the backend respond with no partitions?

Reply:
The backend returns an error when 0 partitions are requested. We should catch this ourselves since we request one less than what the user requested.

Reply:
I had already ensure we always returned a minimum of 1 QueryPartition. I just added the short circuit in golang though, since it is the case we don't really need to request anything.